lib/scholar/cluster/gmm.ex

```diff
@@ -0,0 +1,407 @@
+    reg_covar = opts[:reg_covar]
+
+    {best_params, _} =
+      while {
+              best_params =
+                init_params(key, x, num_gaussians: num_gaussians, reg_covar: reg_covar),
+              {
+                run = 1,
+                max_lower_bound = Nx.Constants.neg_infinity(),
+                x,
+                key,
+                max_iter,
+                tol,
+                reg_covar
+              }
+            },
+            run <= num_runs do
+        {{lower_bound, params}, _} =
+          while {
+                  {
+                    lower_bound = Nx.Constants.neg_infinity(),
+                    params =
+                      init_params(key, x, num_gaussians: num_gaussians, reg_covar: reg_covar)
+                  },
+                  {
+                    iter = 1,
+                    converged = Nx.tensor(false),
+                    x,
+                    key,
+                    max_iter,
+                    tol,
+                    reg_covar
+                  }
+                },
+                iter <= max_iter and not converged do
+            prev_lower_bound = lower_bound
+            {weights, means, _, precisions_cholesky} = params
+            {log_prob_norm, log_resp} = e_step(x, weights, means, precisions_cholesky)
+            params = m_step(x, log_resp, reg_covar)
+            lower_bound = log_prob_norm
+            change = lower_bound - prev_lower_bound
+            converged = Nx.abs(change) < tol
+
+            {
+              {lower_bound, params},
+              {iter + 1, converged, x, key, max_iter, tol, reg_covar}
+            }
+          end
+
+        {max_lower_bound, best_params} =
+          if max_lower_bound < lower_bound do
+            {lower_bound, params}
+          else
+            {max_lower_bound, best_params}
+          end
+
+        {
+          best_params,
+          {run + 1, max_lower_bound, x, key, max_iter, tol, reg_covar}
+        }
+      end
+
+    {weights, means, covariances, precisions_cholesky} = best_params
+
+    %__MODULE__{
+      weights: weights,
+      means: means,
+      covariances: covariances,
+      precisions_cholesky: precisions_cholesky
+    }
+  end
+
+  defnp init_params(key, x, opts) do
+    type = Nx.type(x)
+    {num_samples, _} = Nx.shape(x)
+    num_gaussians = opts[:num_gaussians]
+    reg_covar = opts[:reg_covar]
+
+    k_means_model =
+      Scholar.Cluster.KMeans.fit(x, num_clusters: num_gaussians, num_runs: 1, key: key)
+
+    labels = k_means_model.labels
+    resp = Nx.take(Nx.eye(num_gaussians, type: type), labels)
+
+    {weights, means, covariances, precisions_cholesky} =
+      estimate_gaussian_parameters(x, resp, reg_covar)
+
+    weights = weights / num_samples
+    {weights, means, covariances, precisions_cholesky}
+  end
+
+  defnp e_step(x, weights, means, precisions_cholesky) do
+    {log_prob_norm, log_resp} = estimate_log_prob_resp(x, weights, means, precisions_cholesky)
```

Comment:
what does _resp mean here?

Reply:
`resp` is short for responsibility or responsibilities. `resp[n][k]` measure how much is the k-th Gaussian responsible for generating the n-th observation. Perhaps using `responsibilities` instead of `resp` might be a better idea.

Reply:
Not sure about this as well. Thoughts?

Reply:
If it wasn't intuitive for Paulo, then I think that it is worth changing it to `responsibilities`.